<!DOCTYPE HTML>
<html>

<head>
  <meta charset="utf-8">
</head>
	

<body>
	<script>
  "use strict";

  function ask(question, answer, ok, fail) {
    var result = prompt(question, '');
    if (result.toLowerCase() == answer.toLowerCase()) ok();
    else fail();
  }

  var user = {
    login: 'Василий',
    password: '12345',

    loginOk: function() {
      alert( this.login + ' вошёл в сайт' );
    },

    loginFail: function() {
      alert( this.login + ': ошибка входа' );
    },

    // решение через замыкание
    // Для того, чтобы избежать проблем, можно использовать this. Внутри checkPassword он всегда будет равен текущему объекту, так что скопируем его в переменную, которую назовём self:
    checkPassword: function() {
      var self = this;
      debugger;
      ask("Ваш пароль?", this.password,
        function() {
          self.loginOk();
        },
        function() {
          self.loginFail();  // this здесь не будет работать т.к. внутри внешней функции нет loginOk и loginFail
        }
      );
    } // Теперь всё работает. Анонимные функции достают правильный контекст из замыкания, где он сохранён в переменной self.
  };

  var vasya = user;
  user = null;
  vasya.checkPassword();
	

	</script>
</body>

</html>